### ISSUE TITLE
Psalm v5: `__toString` of some types does not return `exact` return type anymore

### ISSUE THREAD

**Issue by boesing** [vimeo/psalm#7985]
In psalm v4, using `(string) $type` on an instance of `TNumericString` returned `numeric-string`.
As of psalm v5, this now returns `string`.

Is this an expected behavior?

**Comment by orklah** [comment]
This would probably be an effect of https://github.com/vimeo/psalm/pull/7409

But I don't get how it happens. Calling `__toString` on a `TNumericString` targets `Atomic::__toString` which itself calls `$this->getId()` which resolves to `TNumericString::getId()` with `$exact` to `true`. When calling that, it should return `numeric-string`. Are you sure this is the behavior you have?

**Comment by boesing** [comment]
I at least have failing codeception tests regarding this.

Will see if I can debug the reason.

**Comment by boesing** [comment]
It seems that in case of a `Union`, `false` is passed to `Atomic#getId`:

https://github.com/vimeo/psalm/blob/894e4e4c0208b4cd6626edf8e77456edc18c7779/src/Psalm/Type/Union.php#L371
https://github.com/vimeo/psalm/blob/f960d71b7e8dccb0815bce28a22a33ce5410e361/src/Psalm/Type/Union.php#L381

**Comment by boesing** [comment]
I probably just gonna switch to `Atomic#getId` directly instead.
Since the affected commit did not just change the `Union` but also other places, either reverting the commit or moving over to `Atomic#getId` would be a proper solution, right?

I guess you knew what you did there and thus I better not question that 😂

**Comment by orklah** [comment]
Honestly, the whole thing was a mess before :(

__toString was generally used to return a less detailed version of a type but it was not consistent, some function were calling other and then were overriden to call something else entirely. And nothing was documented.

We have a few BC break for Psalm 5 linked to this PR

```
[BC] Atomic::__toString() used to return a string representation of the type that was using double quotes (") to quote literals. This is now using single quotes (') to be more aligned with the rest of the codebase.

[BC] Atomic::__toString() is now final

[BC] Atomic::__toString() now returns a more detailed version of the type (it calls getId() under the hood)

[BC] Atomic::getId() has now a first param $exact. Calling the method with false will return a less detailed version of the type in some cases (similarly to what __toString used to return)
```

To give you a slight idea, here's what happened in 4.x:
TNumericString extends TNonEmptyString which itself extends TString.

Calling __toString on TString returns `string`
Calling __toString on TNonEmptyString  returns `string`
Calling __toString on TNonEmptyString  returns `numeric-string`

Calling getKey on TString returns `string`
Calling getKey on TNonEmptyString  returns `string`
Calling getKey on TNonEmptyString  returns `numeric-string`

Calling getId on TString returns `string`
Calling getId on TNonEmptyString  returns `non-empty-string`
Calling getId on TNonEmptyString  returns `numeric-string`

So I had to make choices and break things in some places. I kept the original idea that __toString should be the one used for debug and that can be less detailed while getId should always be the more detailed but this could break things in some places when relying on the output of __toString
